### PR TITLE
[WIP] Unresolved symlinks

### PR DESCRIPTION
**Pull request by fmeum**

### CODE REVIEW

**Comment by aiuto** [comment]
Is there a bug and design behind this?
@alexjski : you might be interested in this.

**Comment by fmeum** [comment]
This PR combines the changes of three other PRs (https://github.com/bazelbuild/bazel/pull/15781, https://github.com/bazelbuild/bazel/pull/15773, https://github.com/bazelbuild/bazel/pull/15700) that are currently awaiting review. Their ultimate goal is to stabilize `--experimental_allow_unresolved_symlinks` (https://github.com/bazelbuild/bazel/issues/10298).

File: src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java [hunk 1]
```diff
@@ -1537,8 +1538,15 @@ private boolean checkOutputs(
           if (output.isTreeArtifact()) {
             reportOutputTreeArtifactErrors(action, output, reporter, e);
           } else {
-            // Are all exceptions caught due to missing files?
-            reportMissingOutputFile(action, output, reporter, output.getPath().isSymbolicLink(), e);
+            if (e instanceof NotASymlinkException) {
+              reporter.handle(Event.error(
+                  action.getOwner().getLocation(),
+                  String.format("declared output '%s' is not a symlink", output.prettyPrint())));
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java:1544]: How about "Output '%s' was declared as a symlink, but is a '%s'"?

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java [hunk 2]
```diff
@@ -185,14 +185,14 @@ protected PathFragment readSymbolicLink(PathFragment path) throws IOException {
   }
 
   @Override
-  protected void createSymbolicLink(PathFragment linkPath, PathFragment targetFragment)
+  protected void createSymbolicLink(PathFragment linkPath, String rawTarget)
```
Comment (larsrc-google) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:188]: I am none too happy about turning this into just taking a raw string. What is it about the PathFragment normalization that needs to be addressed in this commit? Passing around simple strings increases the ways things can go wrong.

File: src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java (hunk 1 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/skyframe/SkyframeActionExecutor.java:1544]: Good point, will update the corresponding PR if possible.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java (hunk 2 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java:188]: A symlink with target `bar/../foo` is not the same as a symlink with target `foo`: If `bar` is itself a symlink (or a junction on Windows), then `bar/../foo` will resolve relative to the target of `bar`.

Given these complications around symlinks, there has to be a way for Bazel to create "raw" symlinks via its filesystem abstractions. The respective PR is still WIP, but I plan to add comments explaining that the `PathFragment` overload should be preferred in almost all cases - except where it doesn't work, e.g. when explicitly creating a raw symlink.

**Comment by alexjski** [comment]
> Is there a bug and design behind this?

I agree that the context would be really useful here. That also goes together with @larsrc-google's excellent question about using raw string for symlinks--the explanation is sufficient for why that technically is needed, but it would be interesting to know what particular use cases need the symlinks not to be normalized.

**Comment by fmeum** [comment]
@alexjski Please see the individual PRs for context:

* https://github.com/bazelbuild/bazel/pull/15773 fixes https://github.com/bazelbuild/bazel/issues/14224 (discussing the raw vs. normalized symlink issue)
* https://github.com/bazelbuild/bazel/pull/15781 fixes https://github.com/bazelbuild/bazel/issues/10298#issuecomment-1103810996
* https://github.com/bazelbuild/bazel/pull/15700 enables tests on Windows and fixes a bug as explained in the PR description.

More recent context can also be found at https://github.com/bazelbuild/bazel/issues/15454.

The current PR is merely a testbed to verify all the PRs function together, it is not a good starting point for a review (hence is marked as Draft).

**Comment by alexjski** [comment]
> The current PR is merely a testbed to verify all the PRs function together, it is not a good starting point for a review (hence is marked as Draft).

Thank you for the explanation! Should we move the discussion to the corresponding PRs instead?

**Comment by aiuto** [comment]
@alexjski   You might see some context in https://github.com/bazelbuild/rules_pkg/issues/115.

The basic requirement is that when you are building distribution packages, you almost always want symlinks preserved rather than followed.  For example, you want a binary and support files to live under /usr/share/my_package/{bin, lib, messages, ...}, and you want a symlink from `/usr/bin/my_binary` to `/usr/share/my_package/bin/my_binary`.   The issue I pointed to talks about this from the perspective of NodeJS packages where, they have large symlink forests.

I've been approaching this from the rules_pkg side. My latest thinking is that
- Rule authors need to be able to tell if a source file is a symlink or not, so they can offer the ability to the users to either resolve or keep intact
- Some rules might also need the ability to handle symlinks in tree artifacts in one of two ways, but that is not Bazel's problem.  When they walk the tree they can look for symlinks and do what they wan.

**Pull request closed without merging**